Zend/zend_vm_def.h

```diff
@@ -6169,9 +6254,19 @@ ZEND_VM_HANDLER(125, ZEND_FE_RESET_RW, CONST|TMP|VAR|CV, JMP_ADDR)
 			ZEND_VM_NEXT_OPCODE_CHECK_EXCEPTION();
 		} else {
 			zend_class_entry *ce = Z_OBJCE_P(array_ptr);
-			zend_object_iterator *iter = ce->get_iterator(ce, array_ptr, 1);
+			zend_object_iterator *iter;
 			zend_bool is_empty;
 
+			if (UNEXPECTED(ZEND_OBJECT_HAS_TYPE_HINTS(array_ptr))) {
+				zend_throw_exception_ex(
+					zend_ce_type_error, 0,
+					"Typed properties exist in %s: foreach by reference is disallowed",
+					ZSTR_VAL(Z_OBJCE_P(array_ptr)->name));
+				HANDLE_EXCEPTION();
+			}
+
+			iter = ce->get_iterator(ce, array_ptr, 1);
+
```

Comment:
Unfortunately this check is not enough:

<?php
class A {
  public int $x = 5;
}
$a = [0,1];
foreach ($a as &$val) {
  if ($val == 0) {
    $a = new A;
  } else {
    $val = "ops";
  }
}
var_dump($a);'
?>

object(A)#1 (1) {
  ["x"]=>
  &string(3) "ops"
}